app/Middleware/Queue/DriverFactory.php

```diff
@@ -67,6 +69,18 @@ public static function create(array $configuration = [])
                 break;
             case self::DOCTRINE:
                 $connection = DriverManager::getConnection($properties);
+                $config = Config::getInstance();
+                if (!$config->get('CDASH_TESTING_MODE')) {
+                    // Create required tables if they do not already exist.
+                    $tables = ['bernard_messages', 'bernard_queues'];
```

Comment:
Violates SRP. Table creation should not be handled by `DriverFactory`.

Reply:
Fair enough.  Do you have any suggestions on where this initialization step should occur?

Reply:
I think just after you've driver creation is the time to check whether or not your database is configured to use it. Of course, this too will introduce a degree of code smell by the act of having  to check the type, but this seems far less acrimonious than creating database tables in a factory. Let me ask: how tables are added to CDash as they become necessary? Where would be a good place to check the configuration to see if messaging queues are turned on and if so, is the configured driver one that requires database tables?

Reply:
Let me rephrase, I don't think that checking it at driver creation time is ideal. I just think that for the current situation, without having to do too much work, that seems to be the place where we should start.

Reply:
> how tables are added to CDash as they become necessary?

The administrator clicks on the "Upgrade CDash" button on upgrade.php.  We display a warning and a link on viewProjects.php if we detect that the database is out of date.

https://github.com/Kitware/CDash/blob/master/public/api/v1/viewProjects.php#L39

It would obviously be better if this check was performed more generally, but for now it's better than nothing.

> Let me rephrase, I don't think that checking it at driver creation time is ideal. I just think that for the current situation, without having to do too much work, that seems to be the place where we should start.

Agreed.  For now, I'll write a small class called `Middleware/Queue/Initializer` or so and invoke it after `DriverFactory::create()`.